presto-main/src/main/java/com/facebook/presto/sql/planner/RowExpressionInterpreter.java

```diff
@@ -883,8 +883,13 @@ private SpecialCallResult tryHandleLike(CallExpression callExpression, List<Obje
                 if (possibleCompiledPattern == null) {
                     return changed(null);
                 }
-                checkState((resolution.isCastFunction(((CallExpression) possibleCompiledPattern).getFunctionHandle())));
-                possibleCompiledPattern = functionInvoker.invoke(((CallExpression) possibleCompiledPattern).getFunctionHandle(), session, nonCompiledPattern);
+
```

Comment:
I'm looking at the implementation of this function and I'm confused. What's the difference between `nonCompiledPattern` and `value`? It might be simpler to follow the logic if you just re-interpret the argument values unconditionally rather than sometimes do it and some times don't. If there's benefit in using both the processed and unprocessed versions of the arguments, could you document the differences and why this is needed?

Reply:
`value LIKE pattern (escape)`. This is the logic. `value` is for `like` function. `pattern` and `escape` are for `like_pattern` function to compile a pattern (possibly with an escape) to regex. If `like_pattern` is called, then we don't have `nonCompiledPattern`; otherwise, we have `possibleCompiledPattern`.

We cannot always do or don't. It depends on when we wanna serialize it, evaluate it or optimize it. For example, if the given SQL is an expression with all constants, we just evaluate it. Otherwise, we should serialize it on the coordinator and optimize it on the worker. By serializing we mean not to evaluate `like_pattern` because Regex cannot be serialized to be sent to workers. However, on worker, we have to optimize it because we don't want to call `like_pattern` for every row.